src/ExpressReceiver.ts

```diff
@@ -151,58 +150,52 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
   next();
 };
 
-// TODO: this should be imported from another package
-function verifySlackRequest(signingSecret: string): RequestHandler {
-  return async (req , _res, next) => {
-    try {
-      const body: string = (await rawBody(req)).toString();
-      const signature = req.headers['x-slack-signature'] as string;
-      const ts = Number(req.headers['x-slack-request-timestamp']);
-
-      // Divide current date to match Slack ts format
-      // Subtract 5 minutes from current time
-      const fiveMinutesAgo = Math.floor(Date.now() / 1000) - (60 * 5);
-
-      if (ts < fiveMinutesAgo) {
-        const error = errorWithCode(
-          'Slack request signing verification failed. Timestamp is too old.',
-          ErrorCode.ExpressReceiverAuthenticityError,
-        );
-        next(error);
-      }
-
-      const hmac = crypto.createHmac('sha256', signingSecret);
-      const [version, hash] = signature.split('=');
-      hmac.update(`${version}:${ts}:${body}`);
-
-      if (!tsscmp(hash, hmac.digest('hex'))) {
-        const error = errorWithCode(
-          'Slack request signing verification failed. Signature mismatch.',
-          ErrorCode.ExpressReceiverAuthenticityError,
-        );
-        next(error);
-      }
-
-      // Verification passed, assign string body back to request and resume
-      req.body = body;
-      next();
-    } catch (error) {
-      next(error);
-    }
-  };
-}
-
-const parseBody: RequestHandler = (req, _res, next) => {
-  if (req.headers['content-type'] === 'application/x-www-form-urlencoded') {
-    const parsedBody = querystring.parse(req.body);
+/**
+ * Bolt's own Express body-parser.
+ *
+ * This handler parses `req.body` or `req.rawBody`(on Google Could Platform)
+ * and overwrites `req.body` with the parsed JS object.
+ * Following middlewares can expect `req.body` is no long a string value.
+ */
+const parseBody: RequestHandler = async (req, _res, next) => {
+  const stringBody: string = await extractRequestBodyAsString(req);
```

Comment:
why is this being called here instead of just accessing `req.body`? it seems that the `ExpressSignatureVerifier` will call this same function first, and assign a value back to `req.body`.

In fact, I think there are two concepts that we're trying to separate, but they are very coupled: request verification, and body parsing. Since they are so coupled, would it make sense to implement them in one `RequestHandler`? The one handler could call a helper function that knows how to verify the signature given a body `Buffer`, and a signing secret, a current time, and the values from the request header (signature and timestamp). Once the signature is verified, the `Buffer` contents can be trusted, so it's safe to parse the `Buffer` into an object and assign the result to `req.body`. I feel like the intermediate `string` representation isn't useful, and can be eliminated.

Reply:
Thank you for the clear explanation. Now I understand the point, and totally agree with the plan described here. I was also thinking that having temporary string representation is a bit confusing and actually has no benefit. I'm going to merge the two into one `RequestHandler` named `verifySignatureAndParseBody` (let me know if you have a better name for it).